### PR TITLE
Add EntityFramework Localization Sample

### PR DESCRIPTION
**Pull request by hishamco**
I see many users asking about localization sources, and how to use other sources such as json, database .. etc instead of resx files. I provide this example to use EntityFramework as source
/cc @davidfowl

### CODE REVIEW

**Comment by hishamco** [comment]
@Eilon @kirthik can you have a look if you don't mind, because I have seen many devs looking for such PR

**Comment by Eilon** [comment]
Why open the same PR in two repos? This is a dup of https://github.com/aspnet/Localization/pull/149, which is still being evaluated.

**Comment by hishamco** [comment]
I'd like to let you know that I made the PR here after @davidfowl suggestion, and I forget to close the one in the localization repo, so if you think that the localiztion repo is the correct place that's fine

**Comment by davidfowl** [comment]
@Eilon we should close the other one.

**Comment by hishamco** [comment]
I closed the other PR

File: Localization.EntityFramework/EFStringLocalizer.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        public IStringLocalizer WithCulture(CultureInfo culture)
+        {
+            CultureInfo.DefaultThreadCurrentCulture = culture;
+            return new EFStringLocalizer(_db);
+        }
+
+        public IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCultures)
+        {
+            return _db.Resources.Include(r => r.Culture).Where(r => r.Culture.Name == CultureInfo.CurrentCulture.Name)
```
Comment (Eilon) [inline Localization.EntityFramework/EFStringLocalizer.cs]: Recommend starting a new line on every new LINQ statement:

```c#
return _db.Resources
    .Include(r => r.Culture)
    .Where(r => r.Culture.Name == CultureInfo.CurrentCulture.Name)
    .Select(r => new LocalizedString(r.Key, r.Value, true)); // and use a named param for the bool
```

Please do this throughout the code in this PR.

File: Localization.EntityFramework/Models/Culture.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved. 
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System.Collections.Generic;
+
+namespace EFLocalizationSample.Model
```
Comment (Eilon) [inline Localization.EntityFramework/Models/Culture.cs]: The namespace should also be `.Models` (fix in other files as well).

File: Localization.EntityFramework/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+namespace EFLocalizationSample
+{
+    public class Startup
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddSingleton<IStringLocalizerFactory, EFStringLocalizerFactory>();
+        }
+
+        public void Configure(IApplicationBuilder app, IStringLocalizerFactory factory)
```
Comment (Eilon) [inline Localization.EntityFramework/Startup.cs]: Change param from `factory` to something more specific, e.g. `localizerFactory`.

**Comment by hishamco** [comment]
The PR is updated
thanks @Eilon

**Comment by Eilon** [comment]
@hishamco thanks for the update.

@DamianEdwards care to take a look and review?

**Comment by hishamco** [comment]
Any updates regarding this PR?

**Comment by hishamco** [comment]
@DamianEdwards is this looks good for you?

**Comment by Eilon** [comment]
@hishamco we're working on some improvements to localization (@DamianEdwards is going to log some bugs soon) so we'll hold off on this PR until those changes go through. There will need to be some small changes to this PR but I don't think any major changes.

**Comment by hishamco** [comment]
Thanks for clarifications @Eilon

**Comment by hishamco** [comment]
The PR has been updated to react to latest changes of localization & hosting APIs

**Comment by hishamco** [comment]
@DamianEdwards is this PR fine along side with your last PR #83

**Comment by hishamco** [comment]
@Eilon any updates regarding this

**Comment by Eilon** [comment]
@hishamco sorry I was on vacation and was then sick.

Can you please squash the commits into one?

I think aside from that this looks :shipit: and I'll let @DamianEdwards merge it in if he doesn't spot anything else.

**Comment by hishamco** [comment]
No problem @Eilon I hope you are fine now .. I will squash the commits

**Comment by hishamco** [comment]
@Eilon any updates regarding this PR, because i'm planning to push another localization PR

File: Localization.EntityFramework/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            await context.Response.WriteAsync($"    <option value=\"{new CultureInfo("zh-CHT").Name}\">{new CultureInfo("zh-CHT").DisplayName}</option>");
+#endif
+            await context.Response.WriteAsync($"    <option value=\"en-NOTREAL\">English (Not a real locale)</option>");
+            await context.Response.WriteAsync($"    <option value=\"pp-NOTREAL\">Made-up (Not a real anything)</option>");
+        }
+        
+        public static void Main(string[] args)
+        {
+            var application = new WebApplicationBuilder()
+                .UseConfiguration(WebApplicationConfiguration.GetDefault(args))
```
Comment (Eilon) [inline Localization.EntityFramework/Startup.cs]: Could you add a call to `.UseServer("Microsoft.AspNetCore.Server.Kestrel")` here (e.g. see https://github.com/aspnet/Hosting/blob/dev/samples/SampleStartups/StartupFullControl.cs#L17) and delete `hosting.json`?

File: Localization.EntityFramework/project.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        "dnx451": { },
+        "dnxcore50": {
+            "dependencies": {
+                "System.Globalization": "4.0.11-beta-*",
+                "System.Globalization.Extensions": "4.0.1-beta-*"
+            }
+        }
+    },
+
+    "publishExclude": [
```
Comment (Eilon) [inline Localization.EntityFramework/project.json]: Consider removing all these unneeded settings.

File: Localization.EntityFramework/Startup.cs (hunk 4 above)
Comment (hishamco) [inline Localization.EntityFramework/Startup.cs]: Sure

File: Localization.EntityFramework/project.json (hunk 5 above)
Comment (hishamco) [inline Localization.EntityFramework/project.json]: Do you mean "compilationOptions" and "frameworks" sections?

Comment (Eilon) [inline Localization.EntityFramework/project.json]: Just the unneeded publishExclude/exlude items. This project doesn't use node, etc. so there's nothing there to exclude.

**Comment by hishamco** [comment]
The PR has been updated and the commits squashed into single one

**Comment by Eilon** [comment]
@hishamco thanks!

@ryanbrandenburg another one.

**Comment by ryanbrandenburg** [comment]
Shouldn't this be under the sample folder?

**Comment by hishamco** [comment]
Yep, it should be .. but I notice few days back it's outside the samples folder, I will push the change right now

**Comment by hishamco** [comment]
The PR has been updated
thanks @ryanbrandenburg

**Comment by ryanbrandenburg** [comment]
02cdea67

**Pull request closed without merging**